File: Golang/538. Convert BST to Greater Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// func convertBST(root *TreeNode) *TreeNode { // 448ms 又比python慢
//     recursion(root, 0)
//     return root
// }

// func recursion(node *TreeNode, value int) int {
//     if node == nil {
//         return 0
//     } else {
//         sumOfRight := recursion(node.Right, value)
//         sumOfLeft := recursion(node.Left, sumOfRight + value + node.Val)
//         node.Val += sumOfRight + value
//         return node.Val + sumOfLeft - value
//     }
// }

func convertBST(root *TreeNode) *TreeNode { // 上面那个是自己写的，这个抄的GitHub老哥，500ms
    sum := 0
    travel(root, &sum)
    return root
}

// 从大到小遍历 BST 并沿路更新 sum 和 节点值
func travel(root *TreeNode, sum *int) {
    if root == nil {
        return
    }
    travel(root.Right, sum)
    *sum += root.Val
    root.Val= *sum
    travel(root.Left, sum)
}

File: Python3/538. Convert BST to Greater Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        # def recursion(node, value): # 104-112ms
        #     if not node:
        #         return 0
        #     else:
        #         sumright = recursion(node.right, value)
        #         sumleft = recursion(node.left, sumright + value + node.val)
        #         node.val += sumright + value
        #         return node.val + sumleft - value
        # recursion(root, 0)
        # return root
        
        # 另外，由于python没有指针，Go的第二种方法不能实现，只能return回一个值在函数外赋值
        # 尝试使用建立数组[sum]，利用python的引用传递。虽然int型引用传递在此无法发挥作用，但是数组的引用传递应该行
        def recursion(node, sumlist): # 成功，108-116ms 
            if not node: return
            recursion(node.right, sumlist)
            sumlist[0] += node.val
            node.val = sumlist[0]
            recursion(node.left, sumlist)
        sumlist = [0]
        recursion(root, sumlist)
        return root
